### PR TITLE
[ENH] improvements to `_check_soft_dependencies`

### PR DESCRIPTION
**Pull request by fkiraly**
Migrates improvements to `_check_soft_dependencies` from `sktime`:

* https://github.com/sktime/sktime/pull/3925
* https://github.com/sktime/sktime/pull/3955

### CODE REVIEW

File: skbase/testing/utils/_dependencies.py [hunk 1]
```diff
@@ -68,21 +78,35 @@
         raise TypeError(msg)
 
     for package in packages:
+
+        try:
+            req = Requirement(package)
```
[the remark this answers is not in the document]
Comment (RNKuhns) [reply skbase/testing/utils/_dependencies.py:83]: This looks valid. `req` is redefined on line 92 before it is used. I think that maybe the try-except block was meant to replace the code on line 92? Either way, one of the use cases probably needs to be deleted.

**Review by RNKuhns — CHANGES_REQUESTED** [review]
CodeQL item looks legit, so we should fix it.

Also, can we add tests of _check_soft_dependencies?

File: skbase/testing/utils/_dependencies.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fkiraly) [reply skbase/testing/utils/_dependencies.py:83]: well, this proves it doesn't only produce garbage! Indeed this is valid. Fixed.

**Comment by fkiraly** [comment]
> Also, can we add tests of `_check_soft_dependencies`?

Sure! Note, there weren't any to start with, and it is not obvious how to ensure the test condition is constant, because it is always using the environment constructed from `pyproject`.

I can test against that environment (assuming the lower bounds will always stay the same) and some negative cases.

**Comment by fkiraly** [comment]
Have added tests.

**Review by RNKuhns — APPROVED** [review]
Looks reasonable to me now. I'll merge the main branch in to re-run the tests, and then merge once they pass.

**Pull request merged**